extensions/chathistory.md

```diff
@@ -92,6 +91,18 @@ This is useful for retrieving conversation context around a single message.
 
 Request up to `limit` number of messages between the given `timestamp` or `msgid` values. With respect to the limit, the returned messages MUST be counted starting from and excluding the first message selector, while finishing on and excluding the second. This may be forwards or backwards in time.
 
+#### `TARGETS`
+
+Unlike the other subcommands, `TARGETS` does not return message history. Instead, it lists channels the user has visible history in, together with users with which the user has exchanged direct messages, ordered by the time of the latest message in the channel history or direct message conversation (i.e. sent from or to the other user). This allows the client to discover missed direct message conversations with users it is not currently aware of.
+
+    CHATHISTORY TARGETS <timestamp=YYYY-MM-DDThh:mm:ss.sssZ> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ> <limit>
```

Comment:
Most of the time here the "until" timestamp is the current timestamp, would it make sense to allow the client to send a placeholder so it does not need to figure out the server's current time?

```
CHATHISTORY TARGETS <timestamp=YYYY-MM-DDThh:mm:ss.sssZ> * *
```

Reply:
Need to be careful wrt. race conditions.

Reply:
My understanding is that the most common case of this is analogous to `CHATHISTORY LATEST`: list all targets starting from the current time and paginating backwards, possibly until a lower bound is reached (estimated from the time of the last message received on the connection).

Here's what I thought: unlike the other commands, `TARGETS` can only handle timestamps, not message IDs (because the spec is designed for compatibility with systems where there is no global ordering on message IDs). In this context, it makes sense to simplify the specification of the command as much as possible --- a single command, with parameters of a single form.

The intended way to express the current time is as in the pseudocode example: take the local current time and add some constant.

Reply:
My recommendation is to use the `time` tag from the last received message from the server as the upper bound. That way, if another user decides to send us a PRIVMSG while we're still in the process of fetching chat history, we won't get the message twice.

Reply:
> My recommendation is to use the time tag from the last received message from the server as the upper bound (e.g. from RPL_WELCOME).

The welcome can be replayed with older timestamps than the current connection. For example with the ZNC bouncer, the time attached to 001/005/MOTDs numerics can be the time in which the ZNC server received the message from its upstream server. This can be days, months, years before the current server time. I don't see a way I can trust any time message from connection handshake as they are sometimes replayed in parts, and cannot be treated as "current" server time.

For the "until time" in this command, I don't see as a client we'd have anything to put here other than making up a date in the servers future time which isn't anymore useful than the server using its own time upon receiving the message (because at this point its the only part of this that knows the server's time).

Reply:
> For example with the ZNC bouncer, the time attached to 001/005/MOTDs numerics can be the time in which the ZNC server received the message from its upstream server.

Sounds like a ZNC bug to me.